Fix PR testing on Linux

cachemeifyoucan opened the pull request:
Fix code compatibility issue on Linux builders due to implicit move
return value.

cachemeifyoucan committed:
Fix PR testing on Linux

Fix code compatibility issue on Linux builders due to implicit move
return value.

cachemeifyoucan commented:
@swift-ci test

cachemeifyoucan commented:
@swift-ci Please test Linux platform

dexonsmith commented:
> Fix Linux PR failure due to old gcc compatibility
> Clean up TableGen API to avoid using Optional<TreePathPrefixMapper>
> which causes some GCC version to error out on them.

Why does it error out? fc7ce2f53b70e0c46642733a570a46811bbbd35f looks pretty indirect; if there's a constructor or something we can add to TreePathPrefixMapper that might be better.

Also, is this old version of GCC something that llvm.org supports building with, or should we use a newer version?

cachemeifyoucan commented:
> 
The failure is here:
https://ci.swift.org/view/Pull%20Request/job/apple-llvm-project-pr-linux/3144/console

The goal for this PR is to step by step to fix the build on Linux so PR testing will pass. I am trying to do this step by step using the CI but I guess the current version in PR is quite old and it triggers quite some issues one by one.

The PR is using default compiler on Ubuntu 16.04, which is `gcc5.4.0`, which means yes, that is a supported compiler version.

cachemeifyoucan committed:
More gcc compatibility fixes for CAS changes

Fix the build failures when using gcc-5.4

cachemeifyoucan commented:
@swift-ci test

dexonsmith commented:
As I said, I think fc7ce2f is the wrong workaround. We should add a workaround to the TreePathPrefixMapper so it works in Optional rather than changing the APIs.

Error was:
```
19:56:20 In file included from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/STLExtras.h:19:0,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:15,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/TableGen/ScanDependencies.h:12,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/lib/TableGen/ScanDependencies.cpp:9:
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h: In instantiation of 'void llvm::optional_detail::OptionalStorage<T, <anonymous> >::emplace(Args&& ...) [with Args = {const llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator, 4096ul, 4096ul, 128ul>&}; T = llvm::BumpPtrAllocatorImpl<>; bool <anonymous> = false]':
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:73:14:   required from 'constexpr llvm::optional_detail::OptionalStorage<T, <anonymous> >::OptionalStorage(const llvm::optional_detail::OptionalStorage<T, <anonymous> >&) [with T = llvm::BumpPtrAllocatorImpl<>; bool <anonymous> = false]'
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:248:13:   required from 'struct std::is_trivially_move_constructible<llvm::TreePathPrefixMapper>'
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:55:79:   required from 'class llvm::Optional<llvm::TreePathPrefixMapper>'
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/lib/TableGen/ScanDependencies.cpp:219:34:   required from here
19:56:20 
/home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:112:5: error: use of deleted function 'llvm::BumpPtrAllocatorImpl<>::BumpPtrAllocatorImpl(const llvm::BumpPtrAllocatorImpl<>&)'
19:56:20      ::new ((void *)std::addressof(value)) T(std::forward<Args>(args)...);
19:56:20      ^
19:56:20 In file included from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/CAS/ThreadSafeAllocator.h:12:0,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/CAS/FileSystemCache.h:18,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/CAS/CachingOnDiskFileSystem.h:12,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/lib/TableGen/ScanDependencies.cpp:10:
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/Support/Allocator.h:67:7: note: 'llvm::BumpPtrAllocatorImpl<>::BumpPtrAllocatorImpl(const llvm::BumpPtrAllocatorImpl<>&)' is implicitly declared as deleted because 'llvm::BumpPtrAllocatorImpl<>' declares a move constructor or move assignment operator
19:56:20  class BumpPtrAllocatorImpl
19:56:20        ^
```
Looks like the probably is that `BumpPtrAllocatorImpl`'s copy constructor is being called for some reason. May need to adjust the TreePathPrefixMapper constructor to be more explicit in some way.

dexonsmith commented:
Seems related to https://reviews.llvm.org/D93510#2499262

cachemeifyoucan commented:
> As I said, I think [fc7ce2f](https://github.com/apple/llvm-project/commit/fc7ce2f53b70e0c46642733a570a46811bbbd35f) is the wrong workaround. We should add a workaround to the TreePathPrefixMapper so it works in Optional rather than changing the APIs.
> 
> Error was:
> 
> ```
> 19:56:20 In file included from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/STLExtras.h:19:0,
> 19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:15,
> 19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/TableGen/ScanDependencies.h:12,
> 19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/lib/TableGen/ScanDependencies.cpp:9:
> 19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h: In instantiation of 'void llvm::optional_detail::OptionalStorage<T, <anonymous> >::emplace(Args&& ...) [with Args = {const llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator, 4096ul, 4096ul, 128ul>&}; T = llvm::BumpPtrAllocatorImpl<>; bool <anonymous> = false]':
> 19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:73:14:   required from 'constexpr llvm::optional_detail::OptionalStorage<T, <anonymous> >::OptionalStorage(const llvm::optional_detail::OptionalStorage<T, <anonymous> >&) [with T = llvm::BumpPtrAllocatorImpl<>; bool <anonymous> = false]'
> 19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:248:13:   required from 'struct std::is_trivially_move_constructible<llvm::TreePathPrefixMapper>'
> 19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:55:79:   required from 'class llvm::Optional<llvm::TreePathPrefixMapper>'
> 19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/lib/TableGen/ScanDependencies.cpp:219:34:   required from here
> 19:56:20 
> /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:112:5: error: use of deleted function 'llvm::BumpPtrAllocatorImpl<>::BumpPtrAllocatorImpl(const llvm::BumpPtrAllocatorImpl<>&)'
> 19:56:20      ::new ((void *)std::addressof(value)) T(std::forward<Args>(args)...);
> 19:56:20      ^
> 19:56:20 In file included from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/CAS/ThreadSafeAllocator.h:12:0,
> 19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/CAS/FileSystemCache.h:18,
> 19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/CAS/CachingOnDiskFileSystem.h:12,
> 19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/lib/TableGen/ScanDependencies.cpp:10:
> 19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/Support/Allocator.h:67:7: note: 'llvm::BumpPtrAllocatorImpl<>::BumpPtrAllocatorImpl(const llvm::BumpPtrAllocatorImpl<>&)' is implicitly declared as deleted because 'llvm::BumpPtrAllocatorImpl<>' declares a move constructor or move assignment operator
> 19:56:20  class BumpPtrAllocatorImpl
> 19:56:20        ^
> ```
> 
> Looks like the probably is that `BumpPtrAllocatorImpl`'s copy constructor is being called for some reason. May need to adjust the TreePathPrefixMapper constructor to be more explicit in some way.

I think this is a red herring from compiler. There is only an optional<BumpPtrAllocator> in the PrefixMapper which is a member of the TreePathPrefixMapper. Also I like my interface better.

dexonsmith commented:
> Seems related to https://reviews.llvm.org/D93510#2499262

Based on comments in that review, I suspect the wokraround is to change OptionalStorage's detection to use llvm::is_trivially_move_constructible instead of std::is_trivially_move_constructible.

Should be reproducible on main by adding an `Optional<BumpPtrAllocator>` to a unit test.

> I think this is a red herring from compiler.

Of course it is, but we need our types to work with Optional for supported compilers, so we need to fix the actual problem. In this case it looks like a simple patch to Optional.h.

If you don't fix it now, we'll have weird trouble later when we want to use it in an Optional and/or are trying to land things upstream.

> There is only an optional in the PrefixMapper which is a member of the TreePathPrefixMapper. Also I like my interface better.

Sure, maybe it's better, maybe it isn't (we can discuss), but it shouldn't be done in a PR for fixing the Linux build.

dexonsmith commented on llvm/include/llvm/CAS/OnDiskHashMappedTrie.h:
```diff
@@ -57,7 +57,7 @@ class OnDiskHashMappedTrie {
     explicit operator bool() const { return I == -2u; }
 
-    LookupResult() noexcept = default;
+    LookupResult() = default;
```
What's this about? Should this be LLVM_NOEXCEPT?

cachemeifyoucan replied:
It complain that the exception type of the constructor do not match (yes, it is build with -fno-exception). I don't know what is going on.

dexonsmith replied:
Probably a bug related to the `Content` member, since `MappedContentReference` isn't marked `noexcept` (shouldn't matter, but...). Dropping it seems fine; frankly I'm not sure why it was here at all.

cachemeifyoucan replied:
```
/source/llvm/include/llvm/CAS/OnDiskHashMappedTrie.h: In member function 'llvm::cas::OnDiskHashMappedTrie::MappedContentReference llvm::cas::OnDiskHashMappedTrie::insert(llvm::ArrayRef<unsigned char>, llvm::StringRef, llvm::StringRef)':
/source/llvm/include/llvm/CAS/OnDiskHashMappedTrie.h:89:32: error: use of deleted function 'constexpr llvm::cas::OnDiskHashMappedTrie::LookupResult::LookupResult()'
     return insert(LookupResult(), Hash, Metadata, Data);
                                ^
/source/llvm/include/llvm/CAS/OnDiskHashMappedTrie.h:60:5: note: 'constexpr llvm::cas::OnDiskHashMappedTrie::LookupResult::LookupResult() noexcept' is implicitly deleted because its exception-specification does not match the implicit exception-specification ''
     LookupResult() noexcept = default;
     ^
```

dexonsmith commented on llvm/include/llvm/TableGen/ScanDependencies.h:
```diff
@@ -78,11 +79,11 @@ struct ScanIncludesResult {
+Expected<ScanIncludesResult> scanIncludes(
+    cas::CASDB &CAS, cas::CASID ExecID, StringRef MainFilename,
+    ArrayRef<std::string> IncludeDirs, cas::CachingOnDiskFileSystem &FS,
+    function_ref<StringRef(const vfs::CachedDirectoryEntry &)> RemapPath =
```
For fixing the build, please fix Optional or TreePathPrefixMapper rather than changing the API. (If youd on't like the API we should change it separately).

cachemeifyoucan replied:
Experiment a bit here:
* Using `llvm::is_trivially_move_constructible` fixes the problem
* I can't seem to produce a unit test for this as `Optional<BumpPtrAllocator>` works fine and few other simple example I tried. Probably need to have a good and deep understanding of the compiler "bug" to write a test case if I want to fix this upstream. @dexonsmith Any suggestions?

dexonsmith replied:
Great news, glad that works.

For the test, probably `Optional<BumpPtrAllocator>` on its own reduced the testcase too much.
```
struct ImplicitlyNonMovable {
  ImplicitlyNonMovable(); // Constructor out-of-line.
  Optional<BumpPtrAllocator> Alloc;
};
void func() { Optional<ImplicitlyNonMovable> V; }
```
If that doesn't trigger it either, maybe try adding parameters to the constructor, or adding a third layer (since the original code had three):
```
struct T1 {
  T1();
  Optional<BumpPtrAllocator> Alloc;
};
struct T2 {
  T2();
  T1 M1;
};
void func() { Optional<T2> V; }
```
Also, probably better to use a minimal type rather than BumpPtrAllocator in the test for upstream, but you can reduce that later.

BTW, for the purposes of this PR (getting the bots going), I don't think you need a test. Just fix Optional with a FIXME to add a test. (But I know you need one for the upstream patch.)

dexonsmith replied:
Oh, this comment might point at a failing example: https://reviews.llvm.org/D93510#2538983

Looks at least like you could copy the godbolt example and iterate there, if that's not already what you're doing.

cachemeifyoucan replied:
Unfortunately I tried all of those. If I have time later, I might try to creduce a test case.

dexonsmith replied:
Did you try the `virtual ~... = default` suggested in the comment I just linked to?

dexonsmith replied:
Here's a link to a repro https://godbolt.org/z/c4oso1TTe

dexonsmith replied:
This is a more accurate repro https://godbolt.org/z/xbP68YG47

cachemeifyoucan replied:
Ah, that is the piece I am missing. Here you go: https://reviews.llvm.org/D117254

dexonsmith replied:
Manually reduced on godbolt (https://godbolt.org/z/934cbao4T):
```
struct NoCopy {
  NoCopy() = default;
  NoCopy(const NoCopy &) = delete;
};
struct HasNoCopyOptional {
  ~HasNoCopyOptional();
  Optional<NoCopy> Alloc;
};

void func() { Optional<HasNoCopyOptional> O; }
```

dexonsmith replied:
Confirmed:
- Fails in GCC 5.4, unless you switch to `llvm::is_trivially_move_constructible`
- Works in clang
- Works in newer GCC

dexonsmith replied:
Aha, we both got it working at the same time :)

I reviewed; looks good, but I'd be more comfortable with someone from the previous discussion signing off. Can be cherry-picked in here nevertheless.

cachemeifyoucan commented:
Note I tested the linux build with default gcc on 20.04 before and there is no error.

cachemeifyoucan commented:
@swift-ci Please test Linux platform

cachemeifyoucan committed:
[ADT] Fix Optional<> with llvm::is_trivially_move_constructible

Fix the compatibility of Optional<> with some GCC versions that it will fail
to compile when T is getting checked for `is_trivially_move_constructible`
as mentioned here: https://reviews.llvm.org/D93510#2538983

Fix the problem by using `llvm::is_trivially_move_constructible`.

Reviewed By: jplayer-nv, tatyana-krasnukha

Differential Revision: https://reviews.llvm.org/D117254

cachemeifyoucan commented:
@swift-ci test

cachemeifyoucan commented:
@swift-ci test
